clippy_lints/src/methods/mod.rs

```diff
@@ -3284,6 +3285,30 @@ declare_clippy_lint! {
     "calling `.drain(..).collect()` to move all elements into a new collection"
 }
 
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for usage of `Iterator::fold` with a type that implements `Try`.
+    ///
+    /// ### Why is this bad?
+    /// This is better represented with `try_fold`, but this has one major difference: It will
+    /// short-circuit on failure. *This is almost always what you want*. This can also open the door
+    /// for additional optimizations as well, as rustc can guarantee the function is never
+    /// called on `None`, `Err`, etc., alleviating otherwise necessary checks.
```

Comment:
I think the optimization part should go first, as the lint is in the `perf` category. Then, mention legibility, and then in the `### Known issues` mention that maybe the user doesn't *always* want this and it may not be the best solution for these exceptions.